example/particles/particles.cpp

```diff
@@ -162,34 +180,48 @@ TaskStatus DepositParticles(MeshBlock *pmb) {
   auto swarm_d = swarm->GetDeviceContext();
 
   auto &particle_dep = pmb->meshblock_data.Get()->Get("particle_deposition").data;
-  // Reset particle count
-  pmb->par_for(
-      "ZeroParticleDep", kb.s, kb.e, jb.s, jb.e, ib.s, ib.e,
-      KOKKOS_LAMBDA(const int k, const int j, const int i) {
-        particle_dep(k, j, i) = 0.;
-      });
-
   const int ndim = pmb->pmy_mesh->ndim;
 
-  pmb->par_for(
-      "DepositParticles", 0, swarm->GetMaxActiveIndex(), KOKKOS_LAMBDA(const int n) {
-        if (swarm_d.IsActive(n)) {
-          int i = static_cast<int>(std::floor((x(n) - minx_i) / dx_i) + ib.s);
-          int j = 0;
-          if (ndim > 1) {
-            j = static_cast<int>(std::floor((y(n) - minx_j) / dx_j) + jb.s);
-          }
-          int k = 0;
-          if (ndim > 2) {
-            k = static_cast<int>(std::floor((z(n) - minx_k) / dx_k) + kb.s);
-          }
+  if (deposition_method == "per_particle") {
+    // Reset particle count
+    pmb->par_for(
+        "ZeroParticleDep", kb.s, kb.e, jb.s, jb.e, ib.s, ib.e,
+        KOKKOS_LAMBDA(const int k, const int j, const int i) {
+          particle_dep(k, j, i) = 0.;
+        });
+
+    pmb->par_for(
+        "DepositParticles", 0, swarm->GetMaxActiveIndex(), KOKKOS_LAMBDA(const int n) {
+          if (swarm_d.IsActive(n)) {
+            int i = static_cast<int>(std::floor((x(n) - minx_i) / dx_i) + ib.s);
+            int j = 0;
+            if (ndim > 1) {
+              j = static_cast<int>(std::floor((y(n) - minx_j) / dx_j) + jb.s);
+            }
+            int k = 0;
+            if (ndim > 2) {
+              k = static_cast<int>(std::floor((z(n) - minx_k) / dx_k) + kb.s);
+            }
 
-          if (i >= ib.s && i <= ib.e && j >= jb.s && j <= jb.e && k >= kb.s &&
-              k <= kb.e) {
-            Kokkos::atomic_add(&particle_dep(k, j, i), weight(n));
+            if (i >= ib.s && i <= ib.e && j >= jb.s && j <= jb.e && k >= kb.s &&
+                k <= kb.e) {
+              Kokkos::atomic_add(&particle_dep(k, j, i), weight(n));
+            }
           }
-        }
-      });
+        });
+  } else if (deposition_method == "per_cell") {
+    pmb->par_for(
+        "DepositParticlesByCell", kb.s, kb.e, jb.s, jb.e, ib.s, ib.e,
+        KOKKOS_LAMBDA(const int k, const int j, const int i) {
+          particle_dep(k, j, i) = 0.;
+          for (int n = 0; n < swarm_d.GetParticleCountPerCell(k, j, i); n++) {
+            const int idx = swarm_d.GetFullIndex(k, j, i, n);
+            particle_dep(k, j, i) += weight(idx);
+          }
+        });
```

Comment:
Maybe add a comment that this is purely meant as an example of the machinery and that it's not recommended to use this approach to deposit particles in practice from a performance point of view?

Reply:
I had thought that in at least some cases (many particles per cell, for example) this code path might be more performant than the use of atomics, although maybe I am too distrustful of atomic operations. This example can actually be used to answer that performance question for varying numbers of cells and particles. Is your expectation that the atomics approach will always win?

Reply:
Good question. I guess the answer will strongly depend on whether the particles are actually sorted or not.
Probably only benchmarking will tell.